core/crypto/sm4.c

```diff
@@ -0,0 +1,270 @@
+	 0x11, 0xd9, 0x5c, 0x41, 0x1f, 0x10, 0x5a, 0xd8},
+	{0x0a, 0xc1, 0x31, 0x88, 0xa5, 0xcd, 0x7b, 0xbd,
+	 0x2d, 0x74, 0xd0, 0x12, 0xb8, 0xe5, 0xb4, 0xb0},
+	{0x89, 0x69, 0x97, 0x4a, 0x0c, 0x96, 0x77, 0x7e,
+	 0x65, 0xb9, 0xf1, 0x09, 0xc5, 0x6e, 0xc6, 0x84},
+	{0x18, 0xf0, 0x7d, 0xec, 0x3a, 0xdc, 0x4d, 0x20,
+	 0x79, 0xee, 0x5f, 0x3e, 0xd7, 0xcb, 0x39, 0x48}
+};
+
+/* System parameter */
+static const uint32_t FK[4] = {
+	0xa3b1bac6, 0x56aa3350, 0x677d9197, 0xb27022dc
+};
+
+/* Fixed parameter */
+static const uint32_t CK[32] = {
+	0x00070e15, 0x1c232a31, 0x383f464d, 0x545b6269,
+	0x70777e85, 0x8c939aa1, 0xa8afb6bd, 0xc4cbd2d9,
+	0xe0e7eef5, 0xfc030a11, 0x181f262d, 0x343b4249,
+	0x50575e65, 0x6c737a81, 0x888f969d, 0xa4abb2b9,
+	0xc0c7ced5, 0xdce3eaf1, 0xf8ff060d, 0x141b2229,
+	0x30373e45, 0x4c535a61, 0x686f767d, 0x848b9299,
+	0xa0a7aeb5, 0xbcc3cad1, 0xd8dfe6ed, 0xf4fb0209,
+	0x10171e25, 0x2c333a41, 0x484f565d, 0x646b7279
+};
+
+static uint8_t sm4Sbox(uint8_t inch)
+{
+	uint8_t *tab = (uint8_t *)SboxTable;
+
+	return tab[inch];
+}
+
+static uint32_t sm4Lt(uint32_t ka)
+{
+	uint32_t bb = 0;
+	uint8_t a[4];
+	uint8_t b[4];
+
+	PUT_UINT32_BE(ka, a, 0);
+	b[0] = sm4Sbox(a[0]);
+	b[1] = sm4Sbox(a[1]);
+	b[2] = sm4Sbox(a[2]);
+	b[3] = sm4Sbox(a[3]);
+	GET_UINT32_BE(bb, b, 0);
+
+	return bb ^ ROTL(bb, 2) ^ ROTL(bb, 10) ^ ROTL(bb, 18) ^ ROTL(bb, 24);
+}
+
+static uint32_t sm4F(uint32_t x0, uint32_t x1, uint32_t x2, uint32_t x3,
+		     uint32_t rk)
+{
+	return x0 ^ sm4Lt(x1 ^ x2 ^ x3 ^ rk);
+}
+
+static uint32_t sm4CalciRK(uint32_t ka)
+{
+	uint32_t bb = 0;
+	uint8_t a[4];
+	uint8_t b[4];
+
+	PUT_UINT32_BE(ka, a, 0);
+	b[0] = sm4Sbox(a[0]);
+	b[1] = sm4Sbox(a[1]);
+	b[2] = sm4Sbox(a[2]);
+	b[3] = sm4Sbox(a[3]);
+	GET_UINT32_BE(bb, b, 0);
+
+	return bb ^ ROTL(bb, 13) ^ ROTL(bb, 23);
+}
+
+static void sm4_setkey(uint32_t SK[32], const uint8_t key[16])
+{
+	uint32_t MK[4];
+	uint32_t k[36];
+	uint32_t i = 0;
+
+	GET_UINT32_BE(MK[0], key, 0);
+	GET_UINT32_BE(MK[1], key, 4);
+	GET_UINT32_BE(MK[2], key, 8);
+	GET_UINT32_BE(MK[3], key, 12);
+
+	k[0] = MK[0] ^ FK[0];
+	k[1] = MK[1] ^ FK[1];
+	k[2] = MK[2] ^ FK[2];
+	k[3] = MK[3] ^ FK[3];
+
+	for (i = 0; i < 32; i++) {
+		k[i + 4] = k[i] ^ sm4CalciRK(k[i + 1] ^ k[i + 2] ^ k[i + 3] ^
+					     CK[i]);
+		SK[i] = k[i + 4];
+	}
+}
+
+static void sm4_one_round(uint32_t sk[32], const uint8_t input[16],
+			  uint8_t output[16])
+{
+	uint32_t i = 0;
+	uint32_t ulbuf[36];
+
+	memset(ulbuf, 0, sizeof(ulbuf));
+
+	GET_UINT32_BE(ulbuf[0], input, 0);
+	GET_UINT32_BE(ulbuf[1], input, 4);
+	GET_UINT32_BE(ulbuf[2], input, 8);
+	GET_UINT32_BE(ulbuf[3], input, 12);
+
+	for (i = 0; i < 32; i++)
+		ulbuf[i + 4] = sm4F(ulbuf[i], ulbuf[i + 1], ulbuf[i + 2],
+				    ulbuf[i + 3], sk[i]);
+
+	PUT_UINT32_BE(ulbuf[35], output, 0);
+	PUT_UINT32_BE(ulbuf[34], output, 4);
+	PUT_UINT32_BE(ulbuf[33], output, 8);
+	PUT_UINT32_BE(ulbuf[32], output, 12);
+}
+
+void sm4_setkey_enc(struct sm4_context *ctx, const uint8_t key[16])
```

Comment:
`void sm4_setkey_enc(..., const uint8_t key[16])`
it is ok to provide such argument (sized table instead of raw pointer)?

Reply:
It's the same but better I would say... since the compiler can check the size and it also better documents the fact that a 16-byte key is expected.